File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.py
class Solution:
    lettersInDigit = {
        2: 'abc',
        3: 'def',
        4: 'ghi',
        5: 'jkl',
        6: 'mno',
        7: 'pqrs',
        8: 'tuv',
        9: 'wxyz',
    }
    
    def appendCharToArray(self, digit: str, arr: List[str])-> List[str]:
        if not arr:
            return [c for c in self.lettersInDigit[int(digit)]]
        
        res = []
        
        for char in self.lettersInDigit[int(digit)]:
            for word in arr:
                res.append(f'{char}{word}')
        
        return res
    
    def letterCombinations(self, digits: str) -> List[str]:
        result = []
        
        for num in digits[::-1]:
            result = self.appendCharToArray(num, result)
        
        return result
            
            
            
        

File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.js
/**
 * @param {string} digits
 * @return {string[]}
 */

const _map = {
        2: ['a','b','c'],
        3: ['d','e','f'],
        4: ['g','h','i'],
        5: ['j','k','l'],
        6: ['m','n','o'],
        7: ['p','q','r','s'],
        8: ['t','u','v'],
        9: ['w','x','y','z']
    }

var letterCombinations = function(digits) {
    if(!digits.length) return []
    const stack = [..._map[digits[0]]]
    
    while(stack[0].length < digits.length){
        const size = stack.length
        const pointer = stack[0].length
        for(let i = 0; i< size; i++){
            const word = stack.shift()
            _map[digits[pointer]].forEach(c => stack.push(word+c))
        }
    }
    
    return stack
};